go/merkletree2/tree.go

```diff
@@ -1,38 +1,405 @@
 package merkletree2
 
-import "sync"
+import (
+	"bytes"
+	"sort"
+	"sync"
+
+	"github.com/keybase/client/go/logger"
+	"golang.org/x/net/context"
+)
 
 // Tree is the MerkleTree class; it needs an engine and a configuration
 // to run
 type Tree struct {
 	sync.RWMutex
-	//eng StorageEngine not yet implemented/defined
-	cfg Config
+	TreeConfig
+
+	eng StorageEngine
+	log logger.Logger
 }
 
-// ValueConstructor is an interface for constructing values, so that typed
-// values can be pulled out of the Merkle Tree. We are of course assuming
-// that there is only one type of Value at the leaves, which makes sense.
-type ValueConstructor interface {
-	// Construct a new template empty value for the leaf, so that the
-	// Unmarshalling routine has the correct type template.
-	Construct() interface{}
+// NewTree makes a new tree
+func NewTree(c TreeConfig, e StorageEngine, l logger.Logger) *Tree {
+	return &Tree{TreeConfig: c, eng: e, log: l}
 }
 
 // Hash is a byte-array, used to represent a full collision-resistant hash.
 type Hash []byte
 
+// Equal compares two hashes byte by byte
```

Comment:
don't really need these comments

Reply:
my linter complains if I don't comment. Not sure if the one in CI does though